### ISSUE TITLE
Long delay when links from epichrome are sent to Safari

### ISSUE THREAD

**Issue by cmacminn** [dmarmor/epichrome#167]
I use Safari as my default browser and I'm experiencing a LONG delay (5-10 seconds) between clicking a link within an epichrome app and the page opening in Safari. It does eventually work.

Not clear whether the problem is with epichrome, Safari, or something else.

macOS Mojave / 10.14.0
Safari 12.0
Epichrome 2.2.3

**Comment by GrogHills** [comment]
I have a similar issue with the Firefox browser.

**Comment by andresrinivasan** [comment]
Chrome as well. I thought I might try to downgrade Epichrome while this was resolved but I couldn't figure out a way without recreating my apps.

**Comment by msubel** [comment]
I'm facing this issue as well and I believe I was able to narrow it down to Epichrome by simply executing `open 'http://same-url-as-in-epichrome` in the terminal and trying with different default browsers.

When invoked from the terminal, Safari (or whatever other default browser) opens immediately. From Epichrome, aver 4-5 seconds.

Epichrome 2.2.3
Epichrome Helper 1.1.2
macOS 10.14 (Mojave)

**Comment by Felix-N** [comment]
Happens here now as well - using Chrome as main browser. It used to be fast, but not anymore.
Recreating the epichrome GMail App does not help either.

**Comment by bramvandenbulcke** [comment]
I have the same issue. I'm also using Chrome as my main browser. I have different browsers on my system (Chrome Canary, Firefox, Vivaldi, Opera) for testing purposes.

**Comment by whatknows** [comment]
Echoing the above with Chrome here.

**Comment by shoobee** [comment]
Experiencing the same issue as everyone else -- any workarounds so far? Thanks!

**Comment by afsoltys** [comment]
Same issue, Chrome! Anyone figure this out?

**Comment by whatknows** [comment]
Hi friends -- not a solution here, but I was just a bit curious (given msubel saw speedy opening via terminal) if this was really epichrome specific, or if other apps have some similar overhead that terminal just doesn't. 

tl;dr -- as far as I can tell, this really is an epichrome only thing.

I had speedy link opening in:
Word (v16) 
Mendeley (v1.19.3)
Mail
... and some random other apps. All speedy.

Other things I've tried:
Setting epichrome's "If a link doesn't match any rule, open it with" is set to "This app" -- super speedy (go figure)
Turning off "Never redirect domain-relative links that would open in the same tab" -- still slow
Changing my default browser from Chrome to Firefox -- still slow

Interesting thing I noticed regardless of the browser: Epichrome will switch focus to the default browser window very quickly, and then it hangs for a while before it then opens a new tab with the link. Almost like some code correctly identifies the application and switches focus to it, but that the open URL command doesn't come along for 5-20 more seconds.

**Comment by afsoltys** [comment]
Good insights @whatknows. I was also playing around with this and noticed if you disable the Epichrome helper completely, it's speedy, but of course results in the same behavior as "If a link doesn't match any rule, open it with" is set to "This app." 

Seeing as we can prove links sent to the default browser are not delayed, and there is also no delay when opening inside the Epichrome app, it seems like the hangup is with the Epichrome Helper _before_ sending out to the system to handle and send to the browser. 

Maybe there's a loop that created or something and it the Helper does not know what to do? And then it eventually spits it out to the browser once it give up?

**Comment by thiloroe** [comment]
Same problem here. The delay only occurs when "Stop further handling for links sent to default browser" is checked though. However, without this checked, Epichrome first opens a new tab, regardless of any rules.

**Comment by robszumski** [comment]
I am also seeing this issue after upgrading to Mojave. No combination of settings fixes it for me.

**Comment by afsoltys** [comment]
anyone come up with a fix for this yet?

**Comment by andresrinivasan** [comment]
Don't judge me, but my fix was to purchase [Coherence Pro](https://www.bzgapps.com/coherencepro). It has its own issues (e.g. no extension to manage redirects) but it keeps me from weeping when I open links.

**Comment by dmarmor** [comment]
Hi all,

Very sorry fixing this has been so slow, I just have very little time to work on Epichrome these days. But I have been following this issue and finally had a chance to poke at it over the last couple days. The problem appears to be Apple's ever-tightening security rules. I've come up with a fix that works on my system (and seems to also solve a couple other, more minor problems). I'm attaching a beta release here, and would appreciate if folks who've experienced this could test it over the next few days and let me know how it's working.

Thanks!

[epichrome-2.2.4b01.dmg.zip](https://github.com/dmarmor/epichrome/files/2908136/epichrome-2.2.4b01.dmg.zip)

**Comment by cmacminn** [comment]
At first glance, it appears to solve the problem on my end. Will let you
know if I encounter any problems over the next few days. Thanks!

Best, Chris

**Comment by thiloroe** [comment]
This worked for me. Great! Thank you!

Thilo

**Comment by whatknows** [comment]
I've been using for a bit now -- it seems to be working here too.
TYVM!!!!

**Comment by dmarmor** [comment]
Closing this issue, as the update seems to have solved it. Will reopen if it crops up again.
